src/ASM/ASMs3D.C

```diff
@@ -2100,21 +2069,21 @@ bool ASMs3D::integrate (Integrand& integrand,
 
           fe.Navg.resize(p1*p2*p3,true);
           double vol = 0.0;
-          int ip = (((i3-p3)*ng[2]*nel2 + i2-p2)*ng[1]*nel1 + i1-p1)*ng[0];
-          for (int k = 0; k < ng[2]; k++, ip += ng[1]*(nel2-1)*ng[0]*nel1)
-            for (int j = 0; j < ng[1]; j++, ip += ng[0]*(nel1-1))
+          size_t ip = 1;
+          for (int k = 0; k < ng[2]; k++)
+            for (int j = 0; j < ng[1]; j++)
               for (int i = 0; i < ng[0]; i++, ip++)
               {
                 // Fetch basis function derivatives at current integration point
-                SplineUtils::extractBasis(spline[ip],fe.N,dNdu);
+                const BasisFunctionVals& bfs = cache.getVals(iel-1,ip);
 
                 // Compute Jacobian determinant of coordinate mapping
                 // and multiply by weight of current integration point
-                double detJac = utl::Jacobian(Jac,fe.dNdX,Xnod,dNdu,false);
-                double weight = dV*wg[0][i]*wg[1][j]*wg[2][k];
+                double detJac = utl::Jacobian(Jac,fe.dNdX,Xnod,bfs.dNdu,false);
+                double weight = dV*wr[i]*wr[j]*wr[k];
```

Comment:
This looks suspicious, when comparing with old code, with different integration order in different directions (not much used, but still..)

Reply:
nice catch. bad copy-paste